internal/tracinginterceptor/interceptor.go

```diff
@@ -39,54 +45,184 @@ var (
-	Tracer opentracing.Tracer
-	// Transport is the name of the transport, it decides the propagation format and propagation carrier
+	Tracer    opentracing.Tracer
 	Transport string
+	Logger    *zap.Logger
 }
 
 // Interceptor is the tracing interceptor for all RPC types.
-// It handles both tracing observability and context propagation using OpenTracing APIs.
 type Interceptor struct {
+	tracer            opentracing.Tracer
+	transport         string
+	propagationFormat opentracing.BuiltinFormat
+	log               *zap.Logger
+}
+
+// PropagationCarrier is an interface to combine both reader and writer interface
+type PropagationCarrier interface {
+	opentracing.TextMapReader
+	opentracing.TextMapWriter
 }
 
 // New constructs a tracing interceptor with the provided parameter.
 func New(p Params) *Interceptor {
-	return &Interceptor{}
+	m := &Interceptor{
+		tracer:            p.Tracer,
+		transport:         p.Transport,
+		propagationFormat: getPropagationFormat(p.Transport),
+		log:               p.Logger,
+	}
+	if m.tracer == nil {
+		m.tracer = opentracing.GlobalTracer()
+	}
+	return m
 }
 
 // Handle implements interceptor.UnaryInbound
 func (m *Interceptor) Handle(ctx context.Context, req *transport.Request, resw transport.ResponseWriter, h transport.UnaryHandler) error {
-	// TODO: implement
-	panic("implement me")
+	extendedWriter, ok := resw.(transport.ExtendedResponseWriter)
+	if !ok {
+		m.log.Debug("ResponseWriter does not implement ExtendedResponseWriter, proceeding without additional tracing")
+		return h.Handle(ctx, req, resw)
+	}
+
+	parentSpanCtx, _ := m.tracer.Extract(m.propagationFormat, getPropagationCarrier(req.Headers.Items(), req.Transport))
+	extractOpenTracingSpan := &transport.ExtractOpenTracingSpan{
+		ParentSpanContext: parentSpanCtx,
+		Tracer:            m.tracer,
+		TransportName:     req.Transport,
+		StartTime:         time.Now(),
+		ExtraTags:         commonTracingTags,
+	}
+	ctx, span := extractOpenTracingSpan.Do(ctx, req)
+	defer span.Finish()
+	err := h.Handle(ctx, req, resw)
+	return updateSpanWithErrorDetails(span, extendedWriter.IsApplicationError(), extendedWriter.ApplicationErrorMeta(), err)
 }
 
 // Call implements interceptor.UnaryOutbound
 func (m *Interceptor) Call(ctx context.Context, req *transport.Request, out transport.UnaryOutbound) (*transport.Response, error) {
-	// TODO: implement
-	panic("implement me")
+	createOpenTracingSpan := &transport.CreateOpenTracingSpan{
+		Tracer:        m.tracer,
+		TransportName: m.transport,
+		StartTime:     time.Now(),
+		ExtraTags:     commonTracingTags,
+	}
+	ctx, span := createOpenTracingSpan.Do(ctx, req)
+	defer span.Finish()
+
+	tracingHeaders := make(map[string]string)
+	if err := m.tracer.Inject(span.Context(), m.propagationFormat, getPropagationCarrier(tracingHeaders, m.transport)); err != nil {
+		ext.Error.Set(span, true)
+		span.LogFields(log.String("event", "error"), log.String("message", err.Error()))
```

Comment:
Put `log.String("event", "error")` in `tagshelper.go`?  `var logFieldEventError`